sdlib/d/sync/rwlock.d

```diff
@@ -0,0 +1,176 @@
+	 *
+	 * Can be locked for reading by any number of readers. No writers can
+	 * acquire the lock until all readers have unlocked.
+	 */
+	void beginRead() shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(RWLock*) &this).lockReadImpl();
+	}
+
+	void endRead() shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(RWLock*) &this).unlockReadImpl();
+	}
+
+	/**
+	 * Lock for writing.
+	 *
+	 * Locks exclusively for the calling thread.
+	 */
+	void beginWrite() shared {
+		mutex.lock();
+
+		(cast(RWLock*) &this).lockWriteImpl();
+	}
+
+	void endWrite() shared {
+		assert(mutex.isHeld());
+		scope(exit) mutex.unlock();
+
+		(cast(RWLock*) &this).unlockWriteImpl();
+	}
+
+private:
+	bool noWriteLock() {
+		return count < Exclusive;
+	}
+
+	void lockReadImpl() {
+		assert(mutex.isHeld());
+		mutex.waitFor(noWriteLock);
+		++count;
+	}
+
+	bool hasExclusiveLock() {
+		return count == Exclusive;
+	}
+
+	void lockWriteImpl() {
+		assert(mutex.isHeld());
+		// Wait for no other exclusive lock.
+		mutex.waitFor(noWriteLock);
+		count += Exclusive;
+		mutex.waitFor(hasExclusiveLock);
+	}
+
+	void unlockReadImpl() {
+		assert(mutex.isHeld());
+		assert((count & ~Exclusive) > 0);
+		--count;
+	}
+
+	void unlockWriteImpl() {
+		assert(mutex.isHeld());
+		assert(count == Exclusive);
+		count = 0;
+	}
+}
+
+unittest rwlock {
+	import core.stdc.pthread;
+	static struct State {
+		shared RWLock lock;
+		bool finish;
+		bool checkFinishState() {
+			lock.beginRead();
+			scope(exit) lock.endRead();
+			assert((lock.count & ~RWLock.Exclusive) > 0);
+
+			return finish;
+		}
+
+		void waitForFinish() {
+			while (!checkFinishState()) {
+				import sys.posix.sched;
+				sched_yield();
+			}
+		}
+
+		void signalFinish() {
+			lock.beginWrite();
+			assert(lock.mutex.isHeld(), "Mutex not held!");
+			assert(lock.count == RWLock.Exclusive, "Not an exclusive lock!");
+			scope(exit) lock.endWrite();
+			finish = true;
+		}
+	}
+
+	State state;
+	assert(state.lock.count == 0);
+	assert(!state.checkFinishState());
+	assert(state.lock.count == 0);
+	state.lock.beginRead();
+	assert(state.lock.count == 1);
+	assert(!state.lock.mutex.isHeld());
+	state.lock.beginRead();
+	assert(state.lock.count == 2);
+	state.lock.endRead();
+	assert(state.lock.count == 1);
+	pthread_t writer;
+	static void* lockForWrite(void* ctx) {
+		auto state = cast(State*) ctx;
+		state.lock.beginWrite();
+		return cast(void*) cast(size_t) state.lock.count;
+	}
+
+	auto r = pthread_create(&writer, null, lockForWrite, &state);
+	assert(r == 0);
+	while (true) {
+		state.lock.mutex.lock();
+		scope(exit) state.lock.mutex.unlock();
+		if (state.lock.count > 1) {
+			break;
+		}
+
+		import sys.posix.sched;
+		sched_yield();
+	}
+
+	assert(state.lock.count == RWLock.Exclusive + 1);
+	state.lock.endRead();
+	void* ret;
+	pthread_join(writer, &ret);
+	assert(cast(size_t) ret == RWLock.Exclusive);
+	state.lock.endWrite();
+	assert(state.lock.count == 0);
+
+	// thread test
```

Comment:
This seems like yet another test, and it would benefit from being its own test. However, I'm not sure what this test is supposed to check for. It does look like a duplicate of the above, but obfuscated by calling methods on `state`.